content/courses/advanced-ada/parts/data_types/types.rst

```diff
@@ -1483,6 +1488,187 @@ representation clause. We discuss this topic
+       --              universal real type
+
+       N      : constant := 10;
+       --                   ^^
+       --           universal integer type
+
+       N_10   : constant := N * 10;
+       --                   ^^^^^^
+       --                operation on
+       --             universal real type
+
+    begin
+       null;
+    end Show_Universal_Real_Integer;
+
+In this example, the expression :ada:`Pi * 2.0` is computed using universal
+real types, while the expression :ada:`N * 10` is computed using universal
+integer types.
+
+Similarly, for anonymous access types, the equality operator uses universal
+access types for the comparison:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Types.Universal_And_Root_Types.Universal_Access
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Universal_Access is
+       I : aliased Integer;
+       A : access Integer := I'Access;
+       B : access Integer := I'Access;
+    begin
+       if A = B then
+          Put_Line ("A = B");
+       else
+          Put_Line ("A /= B");
+       end if;
+    end Show_Universal_Access;
+
+In this example, both :ada:`A` and :ada:`B` are variables of anonymous access
+types. Because the type isn't a known named type, the equality operation
+:ada:`=` uses the universal access type for the comparison.
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm22:`3.3.2 Number Declarations <3-3-2>`
+    - :arm22:`4.5.2 Relational Operators and Membership Tests <4-5-2>`
+
+
+.. _Adv_Ada_Root_Types:
+
+Root Types
+~~~~~~~~~~
+
+The root types can be found on a level above the universal types. In this
+category, we can find the same numeric types that we have for universal types,
+namely the root real, root integer and root fixed types.
+
+The term *root* is used in the context of type derivation. In fact, the root
+type is first type that we derive all other types from. In other words, if we
+declare an integer range as a new type, that type is derived from the root
+integer type. Similarly, if we declare a new floating-point type, that type is
+derived from the root real type. For example:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Types.Universal_And_Root_Types.Root_Integer_Real
+
+    package Show_Root_Integer_Real is
+
+       type Score is range 0 .. 10;
+       --  Type Score is derived from
+       --  the root integer type.
+
+       type Real_Score is
+         digits 10 range 0.0 .. 10.0;
+       --  Type Real_Score is derived from
+       --  the root real type.
+
+    end Show_Root_Integer_Real;
+
+Here, :ada:`Score` and :ada:`Real_Score` are derived from the root integer and
+real types, respectively. Note that the derivation is always implicit, as we
+cannot write something like
+:ada:`type Score is new Root_Integer range 0 .. 10` or
+:ada:`type Real_Score is new Root_Real digits 10 range 0.0 .. 10.0`.
+
+In contrast, if we derive from a standard floating-point or integer type, we're
```

Comment:
Maybe "standard" should be "existing"?

Reply:
Yes, the phrase was a bit confusing, I've just extended it.